### PR TITLE
Config rework

### PR DESCRIPTION
**Pull request by abukosek**
This is still in heavy WIP.

- [X] Pass 1: Convert configuration cmdline flags into structs, make a global config struct, use that everywhere, add loading from YAML config file.
- [x] Pass 2: Make the test runner and other non-node parts use the new config structs.
- [x] Pass 3: Make everything work.

### CODE REVIEW

**Review by kostko — COMMENTED** [review]
Instead of using a global config, can we pass it explicitly?

**Comment by abukosek** [comment]
> Instead of using a global config, can we pass it explicitly?

Hmm, that seems like a lot of extra work.  I'll first finish the current design and then we can discuss changing it.

A lot of modules need access to config options outside of their module, so we'd have to pass the entire global config struct to all modules anyway.  And currently, only a few modules have some kind of init function where we could pass the config in explicitly, so all the others would have to be refactored to include something like that.  I think the current global struct design is cleaner (load once, use anywhere).

**Comment by gw0** [comment]
*Idea:* In order to still support "re-usable" config files (after the removal of cmdline flags) it would be great to support some variable substitution before the config file gets parsed. Could be as trivial as replacing e.g. `${EXTERNAL_IP}` or `{{ EXTERNAL_IP }}` with value of env variable `EXTERNAL_IP` or the value read from a `.env` file. Or a bit more powerful, like in `docker-compose` (https://docs.docker.com/compose/environment-variables/).

**Comment by abukosek** [comment]
> _Idea:_ In order to still support "re-usable" config files (after the removal of cmdline flags) it would be great to support some variable substitution before the config file gets parsed. Could be as trivial as replacing e.g. `${EXTERNAL_IP}` or `{{ EXTERNAL_IP }}` with value of env variable `EXTERNAL_IP` or the value read from a `.env` file. Or a bit more powerful, like in `docker-compose` (https://docs.docker.com/compose/environment-variables/).

I don't think that the `yaml` package we're using to parse the config file supports this.  This looks like something that should be done with external templating tools and not in oasis-core :)

**Comment by ptrus** [comment]
I think being able to set/override config values via environment variables should be possible, unless there is some hard reasoning not to support it.

One option would be using something like [koanf](https://github.com/knadh/koanf) (we use it in https://github.com/oasisprotocol/emerald-web3-gateway). It makes it really easy to use a `Config` struct (as defined in this PR) and automatically support config fields to be overridden via env variables, see example: https://github.com/oasisprotocol/emerald-web3-gateway/blob/a60756a8f1fb05cecd1d95bf31940a32812c9d8a/conf/config.go#L226-L255  (this is technically also possible to do with viper, but it has some issues like: https://github.com/spf13/viper/issues/368 so I'd suggest avoiding it).

**Comment by gw0** [comment]
> I don't think that the `yaml` package we're using to parse the config file supports this. This looks like something that should be done with external templating tools and not in oasis-core :)

It does not, but setting env variables is convenient and adding support (https://github.com/a8m/envsubst) is as simple as:
```go
import "github.com/a8m/envsubst"
data, err := envsubst.ReadFile("config.yaml")
err = yaml.Unmarshal(data, ...)
```

**Comment by abukosek** [comment]
Alright, I'll add that once I'm done with the test runner :)

**Review by kostko — COMMENTED** [review]
Some initial comments.

File: go/common/config/config.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+// GlobalConfig holds the global configuration options.
+var GlobalConfig Config
+
+// Config is the top-level configuration structure.
+type Config struct {
+	// Oasis node mode (validator, non-validator, compute, keymanager, etc.).
+	Mode string `yaml:"mode"`
+	// Node is a consensus validator.
+	ConsensusValidator bool `yaml:"consensus_validator"`
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/common/config/config.go:31]: No idea why we have a separate option for this, I just translated the flag that was already defined before: :man_shrugging:
https://github.com/oasisprotocol/oasis-core/blob/9e12c44f88a62198d25535c489964a2192f8e98d/go/oasis-node/cmd/common/flags/flags.go#L22-L23

File: go/common/config/util.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	"strings"
+	"unicode"
+
+	"github.com/spf13/cast"
+)
+
+// ParseSizeInBytes converts strings like 1GB or 12 mb into an unsigned integer number of bytes.
+// Note: This function was shamelessly lifted from viper:
+// https://github.com/spf13/viper/blob/master/util.go
+func ParseSizeInBytes(sizeStr string) uint {
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/common/config/util.go:13]: Maybe, I'll have a look.

File: go/consensus/tendermint/config/config.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	ListenAddress string `yaml:"listen_address"`
+
+	// Max number of inbound peers.
+	P2PMaxNumInboundPeers int `yaml:"p2p_max_num_inbound_peers"`
+	// Max number of outbound peers (excluding persistent peers).
+	P2PMaxNumOutboundPeers int `yaml:"p2p_max_num_outbound_peers"`
+	// Rate at which P2P packets can be sent (bytes/s).
+	P2PSendRate int64 `yaml:"p2p_send_rate"`
+	// Rate at which P2P packets can be received (bytes/s).
+	P2PRecvRate int64 `yaml:"p2p_recv_rate"`
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/consensus/tendermint/config/config.go:26]: Yeah, I have a comment about that in the P2P code...  These are just translated from the original code.  I'll change them to unsigned.

File: go/consensus/tendermint/config/config.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+type Config struct {
+	// Tendermint mode (full, archive).
+	Mode string `yaml:"mode"`
+
+	// Tendermint address advertised to other nodes.
+	ExternalAddress string `yaml:"external_address"`
+	// Tendermint listen address.
+	ListenAddress string `yaml:"listen_address"`
+
+	// Max number of inbound peers.
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/consensus/tendermint/config/config.go:19]: Oops, I flattened these a bit, will move them into separate structs (also elsewhere).

File: go/oasis-node/cmd/node/run.go [hunk 5]
```diff
@@ -39,7 +39,7 @@ const (
 )
 
 func Mode() string {
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/oasis-node/cmd/node/run.go:41]: Good idea, I'll do that :)

File: go/consensus/tendermint/config/config.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+import (
+	"fmt"
+	"time"
+)
+
+// Config is the Tendermint configuration structure.
+type Config struct {
+	// Tendermint mode (full, archive).
+	Mode string `yaml:"mode"`
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/consensus/tendermint/config/config.go:12]: Yes, after I move the type definitions for these modes into the config package, I'll change this into a proper type.

File: go/consensus/tendermint/config/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (abukosek) [reply go/consensus/tendermint/config/config.go:26]: Oh, I see why these are signed... because tendermint also uses signed types for these :man_facepalming:
We just pass them directly, so in this case I guess it's better to leave them signed in this config and just check that they're >= 0, like I've done in `Validate()`.

File: go/common/config/util.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (abukosek) [reply go/common/config/util.go]: It doesn't look like the YAML package we're using supports that.

[the remark this answers is not in the document]
Comment (kostko) [reply go/common/config/util.go]: Hm, could you define a custom type that wraps uint64 and defines a custom unmarshaller that handles this?

File: go/consensus/api/mode/mode.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import "fmt"
+
+// Mode is the consensus node mode.
+type Mode string
+
+const (
+	// ModeFull is the name of the full node consensus mode.
+	ModeFull Mode = "full"
+	// ModeArchive is the name of the archive node consensus mode.
+	ModeArchive Mode = "archive"
```
Comment (kostko) [inline go/consensus/api/mode/mode.go:13]: Since we are here, why is this needed in addition to the global node mode? Could it be removed?

File: go/oasis-node/cmd/common/config/config.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+			Level:  make(map[string]string),
+		},
+		Debug: DebugConfig{
+			AllowRoot: false,
+			Rlimit:    0,
+		},
+	}
+}
+
+type DebugConfig struct {
```
Comment (kostko) [inline go/oasis-node/cmd/common/config/config.go:33]: Add missing doc comments.

**Review by kostko — COMMENTED** [review]
Can you also make an overview of what configuration options have moved?

File: go/runtime/registry/config/config.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
+// Package config implements global configuration options.
+package config
+
+import (
+	"fmt"
+	"time"
+)
+
+// RuntimeMode is the runtime mode.
+type RuntimeMode string
```
Comment (kostko) [inline go/runtime/registry/config/config.go:10]: This should now be removed in favor of the top-level node mode which already covers all the needed options.

File: go/runtime/registry/config/config.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+)
+
+// Config is the runtime registry configuration structure.
+type Config struct {
+	// Runtime mode (none, compute, keymanager, client, client-stateless).
+	Mode RuntimeMode `yaml:"mode"`
+
+	// Runtime provisioner to use (mock, unconfined, sandboxed).
+	Provisioner RuntimeProvisioner `yaml:"provisioner"`
+	// Paths to runtime resources (format: <path>,<path>,...).
```
Comment (kostko) [inline go/runtime/registry/config/config.go:135]:
```suggestion
	// Paths to runtime bundles.
```

File: go/worker/common/config/config.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
+// Package config implements global configuration options.
+package config
+
+// Config is the common worker configuration structure.
+type Config struct {
+	// Port to use for incoming gRPC client connections.
```
Comment (kostko) [inline go/worker/common/config/config.go:6]: Is this still used anywhere? AFAIK everything has migrated to P2P?

File: go/runtime/registry/registry.go [hunk 12]
```diff
@@ -41,7 +43,7 @@ var ErrRuntimeHostNotConfigured = errors.New("runtime/registry: runtime host not
 // Registry is the running node's runtime registry interface.
 type Registry interface {
 	// Mode returns the configured behavior of runtime workers on this node.
-	Mode() RuntimeMode
+	Mode() regConfig.RuntimeMode
```
Comment (kostko) [inline go/runtime/registry/registry.go:46]: This should be removed in favor of referencing the node configuration.

File: go/runtime/registry/registry.go [hunk 13]
```diff
@@ -70,7 +72,7 @@ type Runtime interface {
 	ID() common.Namespace
 
 	// Mode returns the configured behavior of runtime workers on this node.
-	Mode() RuntimeMode
+	Mode() regConfig.RuntimeMode
```
Comment (kostko) [inline go/runtime/registry/registry.go:75]: This should be removed in favor of referencing the node configuration.

File: go/runtime/registry/config.go [hunk 14]
```diff
@@ -67,95 +37,10 @@ const (
...
-		return true
-	}
-	return false
-}
-
 // RuntimeConfig is the node runtime configuration.
 type RuntimeConfig struct {
 	// Mode is the runtime mode for this node.
-	Mode RuntimeMode
+	Mode regConfig.RuntimeMode
```
Comment (kostko) [inline go/runtime/registry/config.go:43]: This is probably no longer needed now?

File: go/config/config.go [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+	// ModeValidator is the name of the validator node mode.
+	ModeValidator NodeMode = "validator"
+	// ModeNonValidator is the name of the non-validator node mode.
+	ModeNonValidator NodeMode = "non-validator"
+	// ModeCompute is the name of the compute node mode.
+	ModeCompute NodeMode = "compute"
+	// ModeKeyManager is the name of the key manager node mode.
+	ModeKeyManager NodeMode = "keymanager"
+	// ModeClient is the name of the client node mode.
+	ModeClient NodeMode = "client"
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/config/config.go:36]: No idea, we should ask the person who added this mode :)
To me it looks like we could remove it.

File: go/worker/common/config/config.go (hunk 11 above)
Comment (abukosek) [reply go/worker/common/config/config.go:6]: It is still referenced in `GetNodeAddresses()` in `worker/common/config.go`, but that function does not appear to be used anywhere, so I think it's safe to remove.

File: go/config/config.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (kostko) [reply go/config/config.go:36]: Yeah let's remove it. cc @peternose

[the remark this answers is not in the document]
Comment (kostko) [reply go/config/config.go:36]: Oh to be clear, I would remove `non-validator` and leave `client`.

[the remark this answers is not in the document]
Comment (abukosek) [reply go/config/config.go:36]: Yeah, I was also thinking about removing `non-validator` :)

[the remark this answers is not in the document]
Comment (peternose) [reply go/config/config.go:36]: I agree to remove `non-validator`.

File: go/consensus/api/mode/mode.go (hunk 7 above)
Comment (abukosek) [reply go/consensus/api/mode/mode.go]: I've removed this, but it appears that the archive API test wants to have a node that's both a compute and an archive node...  What is the use case for this?  Do we need to keep this functionality or is it OK to remove?

Comment (kostko) [reply go/consensus/api/mode/mode.go]: Huh, which test wants that? It doesn't make any sense.

Comment (kostko) [reply go/consensus/api/mode/mode.go]: I only see a compute node that is _turned into_ an archive node. So it is not both at the same time.

Comment (abukosek) [reply go/consensus/api/mode/mode.go]: The test code seems to have a notion of a "consensus archive node" and a "runtime archive node" -- but an archive node as defined in the main node mode cannot have runtimes.  So somebody added this notion of a "runtime archive node", but why?

Comment (kostko) [reply go/consensus/api/mode/mode.go]: Every archive node should be a runtime archive node if it has any runtimes defined. cc @ptrus

Comment (ptrus) [reply go/consensus/api/mode/mode.go]: I think that this duplication was done in: https://github.com/oasisprotocol/oasis-core/pull/4964

Since originally, I see only the mode defined under consensus (https://github.com/oasisprotocol/oasis-core/pull/4571).

Feel free to assign me on the issue for fixing/cleaning this up if noone else will.

Comment (kostko) [reply go/consensus/api/mode/mode.go]: If it's not too bad we should try to fix it in this PR.

Comment (abukosek) [reply go/consensus/api/mode/mode.go]: It's OK, I've already fixed this in this PR :)

File: go/oasis-test-runner/oasis/oasis.go [hunk 16]
```diff
@@ -179,15 +190,80 @@ func (n *Node) LoadIdentity() (*identity.Identity, error) {
 	return identity.Load(n.dir.String(), factory)
 }
 
+func (n *Node) AddSeedNodesToConfig() {
```
Comment (kostko) [inline go/oasis-test-runner/oasis/oasis.go:193]: Add documentation comments.

File: go/oasis-test-runner/scenario/e2e/registry_cli.go [hunk 17]
```diff
@@ -442,7 +458,7 @@ func (sc *registryCLIImpl) initNode(childEnv *env.Env, ent *entity.Entity, entDi
 			"--" + cmdRegNode.CfgNodeRuntimeID, testNode.Runtimes[0].ID.String(),
 			"--" + cmdSigner.CfgSigner, fileSigner.SignerName,
 			"--" + cmdSigner.CfgCLISignerDir, entDir,
-			"--" + cmdCommon.CfgDataDir, dataDir,
+			"--" + cmdCommon.CfgConfigFile, cfgFile,
```
Comment (kostko) [inline go/oasis-test-runner/scenario/e2e/registry_cli.go:461]: Why is a config file needed for `registry node init`?

File: go/oasis-test-runner/scenario/e2e/registry_cli.go [hunk 18]
```diff
@@ -338,12 +341,12 @@ func (sc *registryCLIImpl) listNodes(childEnv *env.Env) ([]signature.PublicKey,
...
 
 // isRegistered checks if the given node is registered.
-func (sc *registryCLIImpl) isRegistered(childEnv *env.Env, nodeName, nodeDataDir string) error {
+func (sc *registryCLIImpl) isRegistered(childEnv *env.Env, nodeName, nodeCfgFile string) error {
 	sc.Logger.Info(fmt.Sprintf("checking if node %s is registered", nodeName))
 	args := []string{
 		"registry", "node", "is-registered",
 		"--" + grpc.CfgAddress, "unix:" + sc.Net.Validators()[0].SocketPath(),
-		"--" + cmdCommon.CfgDataDir, nodeDataDir,
+		"--" + cmdCommon.CfgConfigFile, nodeCfgFile,
```
Comment (kostko) [inline go/oasis-test-runner/scenario/e2e/registry_cli.go:349]: Why is a config file needed for `registry node is-registered`?

File: go/p2p/config/config.go [hunk 19]
```diff
@@ -0,0 +1,112 @@
+// Package config implements global configuration options.
+package config
+
+import "time"
+
+// Config is the P2P configuration structure.
+type Config struct {
+	// Port to use for incoming P2P connections.
+	HostPort uint16 `yaml:"port"`
```
Comment (kostko) [inline go/p2p/config/config.go:9]: Why not just `Port`?

File: go/p2p/host.go [hunk 20]
```diff
@@ -102,6 +102,7 @@ func (cfg *HostConfig) Load() error {
 
 // ConnManagerConfig describes a set of settings for a connection manager.
 type ConnManagerConfig struct {
+	// XXX: MinPeers in main config is uint32, but int here -- investigate.
```
Comment (kostko) [inline go/p2p/host.go:105]: Does this need investigation?

File: go/p2p/p2p.go [hunk 21]
```diff
@@ -488,6 +488,7 @@ func (cfg *Config) Load() error {
 
 // GossipSubConfig describes a set of settings for a gossip pubsub.
 type GossipSubConfig struct {
+	// XXX: Main config has int64, but here just int -- investigate.
```
Comment (kostko) [inline go/p2p/p2p.go:491]: Does this need investigation?

File: go/runtime/registry/config/config.go [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+// RuntimeProvisioner is the runtime provisioner.
+type RuntimeProvisioner string
+
+const (
+	// RuntimeProvisionerMock is the name of the mock runtime provisioner.
+	//
+	// Use of this provisioner is only allowed if DebugDontBlameOasis flag is set.
+	RuntimeProvisionerMock = "mock"
```
Comment (kostko) [inline go/runtime/registry/config/config.go:16]: FYI, this is not actually using the `RuntimeProvisioner` type (also below).

File: go/runtime/registry/config/config.go [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+	}
+	return nil
+}
+
+// RuntimeEnvironment is the runtime environment.
+type RuntimeEnvironment string
+
+const (
+	// RuntimeEnvironmentSGX specifies to run the runtime in SGX.
+	RuntimeEnvironmentSGX = "sgx"
```
Comment (kostko) [inline go/runtime/registry/config/config.go:51]: FYI, this is not actually using the `RuntimeEnvironment` type (also below).

File: go/runtime/registry/config/config.go [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+type HistoryPrunerConfig struct {
+	// History pruner strategy.
+	Strategy string `yaml:"strategy"`
+	// History pruning interval.
+	Interval time.Duration `yaml:"interval"`
+	// Number of last rounds to keep.
+	NumKept uint64 `yaml:"num_kept"`
+}
+
+func (c *Config) Validate() error {
```
Comment (kostko) [inline go/runtime/registry/config/config.go:92]: Add documentation comments (also elsewhere).

File: go/runtime/registry/registry.go [hunk 25]
```diff
@@ -599,12 +584,12 @@ func New(ctx context.Context, dataDir string, consensus consensus.Backend, ias i
...
 
-	switch cfg.Mode {
-	case RuntimeModeNone:
+	switch config.GlobalConfig.Mode {
+	case config.ModeValidator, config.ModeSeed:
 		r.logger.Info("runtime support is disabled")
 	default:
 		r.logger.Info("runtime support is enabled",
-			"mode", cfg.Mode,
+			"mode", config.GlobalConfig.Mode,
```
Comment (kostko) [inline go/runtime/registry/registry.go:592]: Given that this is now global, it should probably be logged during node startup and removed from here (this entire `switch`).

File: go/worker/common/config.go [hunk 26]
```diff
@@ -75,32 +32,19 @@ func NewConfig() (*Config, error) {
 	}
 
 	cfg := Config{
-		ClientPort:      uint16(viper.GetInt(CfgClientPort)),
 		SentryAddresses: sentryAddresses,
 		TxPool: txpool.Config{
```
Comment (kostko) [inline go/worker/common/config.go:36]: Should we just use `txpool.Config` directly in the config (could move it in a subpackage if needed)? It doesn't make much sense to have this copying here now that we are using config structures anyway?

File: go/config/config.go [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+
+// Config is the top-level configuration structure.
+type Config struct {
+	// Oasis node mode (validator, non-validator, compute, keymanager, etc.).
+	Mode NodeMode `yaml:"mode"`
+
+	// Node is a consensus validator.
+	// This additional option exists because it is currently possible to
+	// run a node that is simultaneously a validator and a compute node.
+	ConsensusValidator bool `yaml:"consensus_validator"`
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/config/config.go:73]: Do we anticipate other options under the `consensus` hierarchy in the future?  I flattened this because I think it's ugly to have just a single option in a struct (the option itself might as well have been named just `validator`).
I can add an extra struct just for this, but since this is a nice opportunity to make the config options nicer, maybe we can keep it as it is now?

File: go/oasis-test-runner/scenario/e2e/registry_cli.go (hunk 17 above)
Comment (abukosek) [reply go/oasis-test-runner/scenario/e2e/registry_cli.go:461]: It needs the data dir from it and I figured it would be nicer for the users if they didn't have to retype the data dir on the command line, but instead just pass their config file (same for `registry node is-registered`).

File: go/p2p/config/config.go (hunk 19 above)
Comment (abukosek) [reply go/p2p/config/config.go:9]: `Port` makes more sense, `HostPort` is what the original author named it.

File: go/p2p/host.go (hunk 20 above)
Comment (abukosek) [reply go/p2p/host.go:105]: Yes (same for other type mismatches in the new p2p stack).

I imagine that the underlying library we're using has the signed types -- in this case, we should also change the config to use signed types and check that they don't overflow, since it's safer than blindly casting an uint to an int.

File: go/runtime/registry/config/config.go (hunk 22 above)
Comment (abukosek) [reply go/runtime/registry/config/config.go:16]: Yeah, this is a copy&paste of the existing registry code -- I'll fix it :)

File: go/worker/common/config.go (hunk 26 above)
Comment (abukosek) [reply go/worker/common/config.go:36]: Yeah, that makes sense.

File: go/p2p/host.go (hunk 20 above)
Comment (abukosek) [reply go/p2p/host.go]: I've modified this as described above.

File: go/worker/common/config.go (hunk 26 above)
Comment (abukosek) [reply go/worker/common/config.go]: I've changed txpool to use the structure defined in the config.

File: .changelog/5070.cfg.md [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+
+- `consensus.tendermint.mode` in favor of using the global node mode (`mode`).
+
+- `runtime.mode` in favor of using the global node mode (`mode`).
+
+- `worker.client.port` as it is no longer needed.
+
+- `worker.registration.force_register` as it is deprecated.
+
+The following options have been renamed:
```
Comment (kostko) [inline .changelog/5070.cfg.md:22]: Given that there have been a lot of changes, can we also make a migration subcommand? So that we can include that in the upgrade instructions.

File: .changelog/5070.cfg.md [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+
+- `ias.debug.skip_verify` to `ias.debug_skip_verify`.
+
+- `runtime.sandbox.binary` to `runtime.sandbox_binary`.
+
+- `runtime.sgx.loader` to `runtime.sgx_loader`.
+
+- `runtime.history.pruner.*` to `runtime.history_pruner.*`.
+
+- `worker.sentry.addresses` to `workers.sentry_addresses`.
```
Comment (kostko) [inline .changelog/5070.cfg.md:83]: I would change `worker.*` to `runtime.*`.

File: .changelog/5070.cfg.md [hunk 30]
```diff
@@ -0,0 +1,105 @@
...
+
+- `debug.rlimit` to `common.debug.rlimit`.
+
+- `debug.allow_root` to `common.debug.allow_root`.
+
+- `pprof.bind` to `pprof.bind_address`.
+
+- `consensus.validator` to `consensus_validator`.
+
+- `consensus.tendermint.*` to `tendermint.*`.
```
Comment (kostko) [inline .changelog/5070.cfg.md:41]: Could we just drop this `tendermint` thing (especially since it will be renamed upstream) from the config and have everything under `consensus`?

File: .changelog/5070.cfg.md [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+- `log.level` to `common.log.level`.  The log level is now a map of module
+to log level.  Use the `default` module to map the default log level.
+
+- `debug.rlimit` to `common.debug.rlimit`.
+
+- `debug.allow_root` to `common.debug.allow_root`.
+
+- `pprof.bind` to `pprof.bind_address`.
+
+- `consensus.validator` to `consensus_validator`.
```
Comment (kostko) [inline .changelog/5070.cfg.md:39]: Change this back.

File: go/config/config.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (kostko) [reply go/config/config.go:73]: See my other comment, I think we should put everything under `consensus` and drop the `tendermint` thing.

File: .changelog/5070.cfg.md (hunk 28 above)
Comment (abukosek) [reply .changelog/5070.cfg.md:22]: As agreed on Slack, we're going to make this in a separate PR later.

File: .changelog/5070.cfg.md [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+
+- `ias.debug.skip_verify` to `ias.debug_skip_verify`.
+
+- `runtime.sandbox.binary` to `runtime.sandbox_binary`.
+
+- `runtime.sgx.loader` to `runtime.sgx_loader`.
+
+- `runtime.history.pruner.*` to `runtime.history_pruner.*`.
+
+- `worker.sentry.addresses` to `runtime.sentry_addresses`.
```
Comment (kostko) [inline .changelog/5070.cfg.md:81]: Why is this not moved together with other sentry changes? I think it should just go under `sentry.addresses`?

File: .changelog/5070.cfg.md [hunk 33]
```diff
@@ -0,0 +1,103 @@
...
+- `worker.registration.*` to `registration.*`.
+
+- `worker.sentry.*` to `sentry.*`.
+
+- `worker.sentry.control.authorized_pubkey` to
+`sentry.control.authorized_pubkeys`.
+
+- `worker.storage.*` to `storage.*`.
+
+- `worker.storage.public_rpc.enabled` to `worker.storage.public_rpc_enabled`.
```
Comment (kostko) [inline .changelog/5070.cfg.md:98]:
```suggestion
- `worker.storage.public_rpc.enabled` to `storage.public_rpc_enabled`.
```

File: .changelog/5070.cfg.md [hunk 34]
```diff
@@ -0,0 +1,103 @@
...
+
+- `worker.storage.checkpoint_sync.disabled` to
+`worker.storage.checkpoint_sync_disabled`.
```
Comment (kostko) [inline .changelog/5070.cfg.md:101]:
```suggestion
`storage.checkpoint_sync_disabled`.
```

File: .changelog/5070.cfg.md [hunk 35]
```diff
@@ -0,0 +1,103 @@
...
+
+- `worker.storage.checkpointer.*` to `storage.checkpointer.*`.
```
Comment (kostko) [inline .changelog/5070.cfg.md:103]: This is probably covered by `worker.storage.*` to `storage.*`?

File: .changelog/5070.cfg.md [hunk 36]
```diff
@@ -0,0 +1,103 @@
...
+- `consensus.tendermint.checkpointer.*` to `consensus.abci.checkpointer.*`.
+
+- `consensus.tendermint.sentry.upstream_address` to
+`consensus.sentry_upstream_addresses`.
+
+- `consensus.tendermint.upgrade.stop_delay` to
+`consensus.upgrade_stop_delay`.
+
+- `consensus.tendermint.supplementarysanity.*` to
+`consensus.supplementary_sanity.`.
```
Comment (kostko) [inline .changelog/5070.cfg.md:63]:
```suggestion
`consensus.supplementary_sanity.*`.
```

File: go/worker/registration/worker.go [hunk 37]
```diff
@@ -1128,15 +1112,9 @@ func New(
...
-	if viper.GetBool(CfgRegistrationForceRegister) {
-		if err = SetForcedDeregister(w.store, false); err != nil {
-			return nil, fmt.Errorf("failed to clear persisted forced-deregister: %w", err)
-		}
-		storedDeregister = false
-	}
 	w.storedDeregister = storedDeregister
 
-	if flags.ConsensusValidator() {
+	if config.GlobalConfig.Consensus.Validator {
```
Comment (kostko) [inline go/worker/registration/worker.go:1117]: This should also work if you set the global mode to validator?

File: go/consensus/tendermint/config/config.go [hunk 38]
```diff
@@ -0,0 +1,228 @@
...
+	ABCI ABCIConfig `yaml:"abci,omitempty"`
+
+	// Consensus state sync configuration.
+	StateSync StateSyncConfig `yaml:"state_sync,omitempty"`
+
+	// Supplementary sanity checks configuration.
+	SupplementarySanity SupplementarySanityConfig `yaml:"supplementary_sanity,omitempty"`
+
+	// Enable tendermint debug logs (very verbose).
+	LogDebug bool `yaml:"log_debug,omitempty"`
```
Comment (kostko) [inline go/consensus/tendermint/config/config.go:48]: Should this go under the debug options?

File: go/oasis-node/cmd/node/node.go [hunk 39]
```diff
@@ -453,6 +454,7 @@ func NewNode() (node *Node, err error) { // nolint: gocyclo
 	// binary is from the logs.
 	logger.Info("Starting oasis-node",
 		"Version", version.SoftwareVersion,
+		"Mode", config.GlobalConfig.Mode,
```
Comment (kostko) [inline go/oasis-node/cmd/node/node.go:457]:
```suggestion
		"version", version.SoftwareVersion,
		"mode", config.GlobalConfig.Mode,
```

File: go/runtime/txpool/config/config.go [hunk 40]
```diff
@@ -0,0 +1,18 @@
+// Package config implements the txpool configuration options.
+package config
+
+import "time"
+
+// TxPoolConfig is the common worker tx pool configuration structure.
+type TxPoolConfig struct {
```
Comment (kostko) [inline go/runtime/txpool/config/config.go:7]: Why not just `Config`?

File: go/runtime/txpool/config/config.go [hunk 41]
```diff
@@ -0,0 +1,18 @@
+// Package config implements the txpool configuration options.
+package config
+
+import "time"
+
+// TxPoolConfig is the common worker tx pool configuration structure.
```
Comment (kostko) [inline go/runtime/txpool/config/config.go:6]:
```suggestion
// TxPoolConfig is the runtime transaction pool configuration structure.
```

File: go/worker/registration/worker.go (hunk 37 above)
Comment (abukosek) [reply go/worker/registration/worker.go:1117]: The previous code only checked the flag:
https://github.com/oasisprotocol/oasis-core/blob/master/go/oasis-node/cmd/common/flags/flags.go#L75-L77

File: go/consensus/tendermint/config/config.go (hunk 38 above)
Comment (abukosek) [reply go/consensus/tendermint/config/config.go:48]: Hmm, it kind of makes sense to leave it here because it only sets the Tendermint log level to debug and isn't some unsupported debug flag.

File: go/worker/registration/worker.go (hunk 37 above)
Comment (kostko) [reply go/worker/registration/worker.go:1117]: Yeah this was wrong, we should make sure that if you set the global mode to validator your node will actually become a validator :-)

**Review by kostko — APPROVED** [review]
